recipes-extended/xen/files/libxl-iso-hotswap.patch

```diff
@@ -0,0 +1,346 @@
++
++    type = libxl__device_disk_string_of_format(format);
++    tap_ctl_find(type, disk, &tap);
++    return tap.minor;
++}
++
+ char *libxl__blktap_devpath(libxl__gc *gc,
+                             const char *disk,
+                             libxl_disk_format format,
+@@ -62,7 +74,7 @@ char *libxl__blktap_devpath(libxl__gc *g
+     return NULL;
+ }
+ 
+-static bool tapdev_is_shared(libxl__gc *gc, const char *params)
++bool tapdev_is_shared(libxl__gc *gc, const char *params)
+ {
+     char **domids, **vbds;
+     char *tp;
+Index: xen-4.6.4/tools/libxl/libxl_internal.h
+===================================================================
+--- xen-4.6.4.orig/tools/libxl/libxl_internal.h
++++ xen-4.6.4/tools/libxl/libxl_internal.h
+@@ -1696,6 +1696,9 @@ _hidden char *libxl__blktap_devpath(libx
+                                     libxl_disk_format format,
+                                     char *keydir);
+ 
++_hidden int libxl__get_tap_minor(libxl__gc *gc, libxl_disk_format format, const char *disk);
++_hidden bool tapdev_is_shared(libxl__gc *gc, const char *params);
++
+ /* libxl__device_destroy_tapdisk:
+  *   Destroys any tapdisk process associated with the backend represented
+  *   by be_path.
+@@ -1762,6 +1765,7 @@ _hidden int libxl__qmp_save(libxl__gc *g
+ /* Set dirty bitmap logging status */
+ _hidden int libxl__qmp_set_global_dirty_log(libxl__gc *gc, int domid, bool enable);
+ _hidden int libxl__qmp_insert_cdrom(libxl__gc *gc, int domid, const libxl_device_disk *disk);
++_hidden int libxl__qmp_change_cdrom(libxl__gc *gc, int domid, const libxl_device_disk *disk);
+ /* Add a virtual CPU */
+ _hidden int libxl__qmp_cpu_add(libxl__gc *gc, int domid, int index);
+ /* close and free the QMP handler */
+Index: xen-4.6.4/tools/libxl/libxl_qmp.c
+===================================================================
+--- xen-4.6.4.orig/tools/libxl/libxl_qmp.c
++++ xen-4.6.4/tools/libxl/libxl_qmp.c
+@@ -948,6 +948,17 @@ int libxl__qmp_set_global_dirty_log(libx
+                            NULL, NULL);
+ }
+ 
++int libxl__qmp_change_cdrom(libxl__gc *gc, int domid,
++							const libxl_device_disk *disk)
++{
++    libxl__json_object *args = NULL;
++    int dev_number = libxl__device_disk_dev_number(disk->vdev, NULL, NULL);
++
++    QMP_PARAMETERS_SPRINTF(&args, "device", "ide-%i", dev_number);
++    qmp_parameters_add_string(gc, &args, "target", "/dev/xvdc");
++    return qmp_run_command(gc, domid, "change", args, NULL, NULL);
++}
++
+ int libxl__qmp_insert_cdrom(libxl__gc *gc, int domid,
+                             const libxl_device_disk *disk)
+ {
+Index: xen-4.6.4/tools/libxl/libxl_utils.c
+===================================================================
+--- xen-4.6.4.orig/tools/libxl/libxl_utils.c
++++ xen-4.6.4/tools/libxl/libxl_utils.c
+@@ -1384,6 +1384,12 @@ int libxl__random_bytes(libxl__gc *gc, u
+     return ret;
+ }
+ 
++int libxl_util_xs_read(libxl_ctx *ctx, char *path, char **out)
```

Comment:
xl_cmdimpl.c doesn't include libxl_internal.h (where the libxl__xs_<op>() functions are declared) and it can't without significant refactoring due to header import restrictions and conflicts.  xl_cmdimpl.c _does_ include libxl_utils.h, which is not a bad place to implement a simple xs_read call.

However it doesn't look like anyone's cleaning up the pointer, so I'll need to add a free in there.